system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -372,6 +393,396 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
         return (findVariable(name, ns_uri, scope)!=nullptr);
     }
 
+    void setLocation(xmlNodePtr node)
+    {
+        WriteLockBlock wblock(m_rwlock);
+        m_xpathContext->doc = node->doc;
+        m_xpathContext->node = node;
+        m_xpathContext->contextSize = 0;
+        m_xpathContext->proximityPosition = 0;
+    }
+
+    bool setLocation(xmlXPathObjectPtr obj, bool required, const char *xpath)
+    {
+        if (!obj)
+        {
+            if (required)
+                throw MakeStringException(XPATHERR_InvalidInput,"XpathContext:setLocation: Error: Syntax error XPATH '%s'", xpath);
+            return false;
+        }
+        if (obj->type!=XPATH_NODESET || !obj->nodesetval || obj->nodesetval->nodeNr==0)
+        {
+            xmlXPathFreeObject(obj);
+            if (required)
+                throw MakeStringException(XPATHERR_EvaluationFailed,"XpathContext:setLocation: Error: Could not evaluate XPATH '%s'", xpath);
+            return false;
+        }
+        if (obj->nodesetval->nodeNr>1)
+        {
+            xmlXPathFreeObject(obj);
+            if (required)
+                throw MakeStringException(XPATHERR_EvaluationFailed,"XpathContext:setLocation: Error: ambiguous XPATH '%s'", xpath);
+            return false;
+        }
+        xmlNodePtr location = obj->nodesetval->nodeTab[0];
+        xmlXPathFreeObject(obj);
+        setLocation(location);
+        return true;
+    }
+
+    virtual bool setLocation(ICompiledXpath * compiledXpath, bool required) override
+    {
+        if (!compiledXpath)
+            return false;
+
+        CLibCompiledXpath * ccXpath = static_cast<CLibCompiledXpath *>(compiledXpath);
+        if (!ccXpath)
+            throw MakeStringException(XPATHERR_InvalidState,"XpathProcessor:evaluateAsNumber: Error: BAD compiled XPATH");
+
+        xmlXPathObjectPtr obj = evaluate(ccXpath->getCompiledXPathExpression(), compiledXpath->getXpath());
+        return setLocation(obj, required, compiledXpath->getXpath());
+    }
+
+    virtual bool setLocation(const char *s, bool required) override
+    {
+        StringBuffer xpath;
+        replaceVariables(xpath, s, false, this, "{$", "}");
```

Comment:
So this change allows us to use the ${...} variable substitution as strings in xpaths wherever this function is used?

Reply:
Yes let me know when you'd like to discuss it.  I'll have to refresh my memory on exactly where this type of substitution is supported.